templates/endomorphism/batchnormalization.wgsl

```diff
@@ -1,14 +1,47 @@
 
 {%- include "structs.wgsl" -%}
 
+[[block]]
+struct Block {
+    data: [[stride({{elem_stride}})]] array<{{ elem_type }}>;
+}; 
+
+// X (input)
 [[group(0), binding(0)]]
-var<storage, read> {{ inputs[0] }}: ArrayVector;
+var<storage, read> {{ inputs[0] }}: Block;
 
+// Scale
 [[group(0), binding(1)]]
-var<storage, write> {{ outputs[0] }}: ArrayVector;
+var<storage, read> {{ inputs[1] }}: Array;
+
+// B (bias)
+[[group(0), binding(2)]]
+var<storage, read> {{ inputs[2] }}: Array;
+
+// Input mean
+[[group(0), binding(3)]]
+var<storage, read> {{ inputs[3] }}: Array;
+
+// Input variance
+[[group(1), binding(0)]]
+var<storage, read> {{ inputs[4] }}: Array;
+
+// Y (Output)
+[[group(1), binding(1)]]
+var<storage, write> {{ outputs[0] }}: Block;
 
 [[stage(compute), workgroup_size(1)]]
 fn main([[builtin(global_invocation_id)]] global_id: vec3<u32>) {
-    let gidx = global_id.x;
-    {{ outputs[0] }}.data[gidx] = ({{ inputs[0] }}.data[gidx] - {{ inputs[3] }}.data[gidx]) / sqrt({{ inputs[4] }}.data[gidx] * {{ inputs[1] }}.data[gidx] + {{ inputs[2] }}.data[gidx])
+    let channel = global_id.y;
+    let batch = global_id.z;
```

Comment:
So be aware that global_id.z has a limited max size of 64: https://www.w3.org/TR/webgpu/#dom-supported-limits-maxcomputeworkgroupsizez

Scaling with threads is easier, but it tends to be limiting for more than 3d tensors

Reply:
Large values should not really be an issue: the number of channels is typically 1...5 and most usually 3, so 64 should be enough (I realize this sounds a lot like '640k will be enough forever!' 🤡 ). Batch size can be larger (depending on your GPU memory of course). If the `z` coordinate is more limited than `x` and `y` I suggest using `y` for batch size and `z` for channels.

Reply:
Well, if batchnorm happens after a convolution the number of channel is going to be the number of convolution kernels and could also be pretty big.

```suggestion
	let gidx = global_id.x;

	let batch = gidx / {{ o_chunks[0][0] }}u; 
	let rest = gidx % {{ o_chunks[0][0] }}u; 

        let channel = rest / {{ o_chunks[0][1] }}u;
        let rest = rest % {{ o_chunks[0][1] }}u;
                
        let y = rest / {{ o_chunks[0][2] }}u;
        let x = rest % {{ o_chunks[0][2] }}u;
```

is going to give you the batch number and channel number if the number of work group on x ( more accurate  term for threads) is equal to the product batch * channel * h * w.

Reply:
Hm this makes a bit more sense. How would you like me to proceed?

Reply:
I am not sure by the way what the limit for x is. You were referring to these, correct? 

<img width="809" alt="image" src="https://user-images.githubusercontent.com/181337/148660553-7984def4-07bc-4299-aac7-bb2563959c19.png">

Says here it is 256, but that refers to the *workgroup size*. In the code you check for x<16352.

I believe the appropriate limit is this one:

<img width="790" alt="image" src="https://user-images.githubusercontent.com/181337/148660602-958fb16e-f29e-4ad0-94b9-7f3898ce4202.png">

So just 65535 for each argument then? 🤔

Reply:
Oh sorry, my bad. Yes, I confiused work group z max size and the dispatch z. which has no more limit than the product of x * y * z max of 65535